### PR TITLE
[SPARK-23214][SQL] cached data should not carry extra hint info

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is a regression introduced by https://github.com/apache/spark/pull/19864

When we lookup cache, we should not carry the hint info, as this cache entry might be added by a plan having hint info, while the input plan for this lookup may not have hint info, or have different hint info.

## How was this patch tested?

a new test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 1]
```diff
@@ -169,14 +169,17 @@ class CacheManager extends Logging {
   /** Replaces segments of the given logical plan with cached versions where possible. */
   def useCachedData(plan: LogicalPlan): LogicalPlan = {
     val newPlan = plan transformDown {
+      // Do not lookup the cache by hint node. Hint node is special, we should ignore it when
+      // canonicalizing plans, so that plans which are same except hint can hit the same cache.
+      // However, we also want to keep the hint info after cache lookup. Here we skip the hint
+      // node, so that the returned caching plan won't replace the hint node and drop the hint info
+      // from the original plan.
+      case hint: ResolvedHint => hint
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: A small clean up for https://github.com/apache/spark/pull/20365

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 2]
```diff
@@ -62,7 +62,7 @@ case class InMemoryRelation(
     @transient child: SparkPlan,
     tableName: Option[String])(
     @transient var _cachedColumnBuffers: RDD[CachedBatch] = null,
-    val batchStats: LongAccumulator = child.sqlContext.sparkContext.longAccumulator,
+    val sizeInBytesStats: LongAccumulator = child.sqlContext.sparkContext.longAccumulator,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: unrelated but this name is more accurate.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala [hunk 3]
```diff
@@ -70,8 +71,8 @@ class BroadcastJoinSuite extends QueryTest with SQLTestUtils {
   private def testBroadcastJoin[T: ClassTag](
       joinType: String,
       forceBroadcast: Boolean = false): SparkPlan = {
-    val df1 = spark.createDataFrame(Seq((1, "4"), (2, "2"))).toDF("key", "value")
-    val df2 = spark.createDataFrame(Seq((1, "1"), (2, "2"))).toDF("key", "value")
+    val df1 = Seq((1, "4"), (2, "2")).toDF("key", "value")
+    val df2 = Seq((1, "1"), (2, "2")).toDF("key", "value")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala]: some code style fixing

**Comment by cloud-fan** [comment]
cc @CodingCat @gatorsmile @hvanhovell @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 4]
```diff
@@ -62,7 +62,7 @@ case class InMemoryRelation(
...
     statsOfPlanToCache: Statistics = null)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: remove `null`

**Comment by CodingCat** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 5]
```diff
@@ -169,14 +169,17 @@ class CacheManager extends Logging {
...
-        lookupCachedData(currentFragment).map { cached =>
-          val cachedPlan = cached.cachedRepresentation.withOutput(currentFragment.output)
-          currentFragment match {
-            case hint: ResolvedHint => ResolvedHint(cachedPlan, hint.hints)
-            case _ => cachedPlan
-          }
-        }.getOrElse(currentFragment)
+        lookupCachedData(currentFragment)
+          .map(_.cachedRepresentation.withOutput(currentFragment.output))
+          .getOrElse(currentFragment)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: Thank you for pinging me, @cloud-fan .
I see. So, this is reverted back from #20365 .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 6]
```diff
@@ -73,11 +73,16 @@ case class InMemoryRelation(
...
-      // Underlying columnar RDD hasn't been materialized, use the stats from the plan to cache
-      statsOfPlanToCache
+    if (sizeInBytesStats.value == 0L) {
+      // Underlying columnar RDD hasn't been materialized, use the stats from the plan to cache.
+      // Note that we should drop the hint info here. We may cache a plan whose root node is a hint
+      // node. When we lookup the cache with a semantically same plan without hint info, the plan
+      // returned by cache lookup should not have hint info. If we lookup the cache with a
+      // semantically same plan with a different hint info, `CacheManager.useCachedData` will take
+      // care of it and retain the hint info in the lookup input plan.
+      statsOfPlanToCache.copy(hints = HintInfo())
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: I am not sure I agree with this. If we cache a plan with a hint, then it is reasonable to expect that the hint is still in the plan. We do the same with temporary views.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: This is a new behavior we introduced in 2.3. I will first keep the behavior unchanged and merge it to 2.3. 

We can have more discussion in the next release.

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master/2.3

**Pull request closed without merging**